railties/lib/rails/tasks/framework.rake

```diff
@@ -66,3 +68,24 @@ namespace :rails do
     end
   end
 end
+
+namespace :rails do
+  Rake.application.top_level_tasks.grep(/^rails:/).each do |task|
+    ActiveSupport::Deprecation.warn(<<-MSG.squish)
+      Running this task with the rails task namespace is deprecated in favor of app.
+      Please run e.g. bin/rails #{task.sub(/^rails:/, 'app:')} instead."
+    MSG
+  end
```

Comment:
The code

``` ruby
Rake.application.top_level_tasks
```

returns only task names given in a command given. (e.g. when `rails rails:update` is given, only `[rails:update]` is returned, not all tasks in the `rails:` namespace.)

So, I think the below snippet is correct.

``` ruby
namespace :rails do
  task :deprecated do |task|
    Rake.application.top_level_tasks.grep(/^rails:/).each do |task|
      ActiveSupport::Deprecation.warn(<<-MSG.squish)
        Running #{task} with the rails: namespace is deprecated in favor of app.
        Run e.g. bin/rails #{task.sub(/^rails:/, 'app:')} instead."
      MSG
    end
  end

  task :update => ['deprecated', 'app:update']
  task :template => ['deprecated',  'app:template']

  namespace :templates do
    task :copy => ['deprecated', 'app:templates:copy']
  end

  namespace :update do
    task :configs => ['deprecated', 'app:update:configs']
    task :bin => ['deprecated', 'app:update:bin']
  end
end
```